Introduce conditional instructions

crisbeto opened the pull request:
Adds the logic to generate instructions for `if` and `switch` blocks. Split up into the following commits to make it easier to review:

### refactor(compiler): generate switch block instructions
Adds the logic to generate the instructions for `switch` instructions. For the following block:

```html
{#switch value()}
  {:case 0} case 0
  {:case 1} case 1
  {:case 2} case 2
  {:default} default
{/switch}
```

The compiler will produce the following output:

```ts
function App_Template(rf, ctx) {
  if (rf & 1) {
    ɵɵtemplate(0, App_Case_0_Template, 1, 0);
    ɵɵtemplate(1, App_Case_1_Template, 1, 0);
    ɵɵtemplate(2, App_Case_2_Template, 1, 0);
    ɵɵtemplate(3, App_Case_3_Template, 1, 0);
  }
  if (rf & 2) {
    let App_contFlowTmp;
    ɵɵconditional(0, (App_contFlowTmp = ctx.value()) === 0 ? 0 : App_contFlowTmp === 1 ? 1 : App_contFlowTmp === 2 ? 2 : 3);
  }
}
```

### refactor(compiler): generate if block instructions
Adds the logic to generate the instructions for `if` blocks. There are two primary use cases we need to account for:

A conditional that doesn't use the `as` parameter of the `if` block. To support it we generate a nested ternary expression that evaluates to the index of the template whose condition is truthy. If the block doesn't have an `else` branch, we pass in a special `-1` value which means that no view will be rendered.

Example with an `else`:
```ts
// {#if expr}
//   ...
//   {:else if otherExpr} ...
//   {:else} ...
// {/if}

if (rf & 1) {
  ɵɵtemplate(0, App_Conditional_0_Template, 0, 0);
  ɵɵtemplate(1, App_Conditional_1_Template, 0, 0);
  ɵɵtemplate(2, App_Conditional_2_Template, 0, 0);
}
if (rf & 2) {
  ɵɵconditional(0, ctx.expr ? 0 : ctx.otherExpr ? 1 : 2);
}
```

Example without an `else`:
```ts
// {#if expr}
//   ...
//   {:else if otherExpr} ...
// {/if}

if (rf & 1) {
  ɵɵtemplate(0, App_Conditional_0_Template, 0, 0);
  ɵɵtemplate(1, App_Conditional_1_Template, 0, 0);
}
if (rf & 2) {
  ɵɵconditional(0, ctx.expr ? 0 : ctx.otherExpr ? 1 : -1);
}
```

If a conditional captures it's value in an alias (e.g. `{#if expr; as foo}`) we need to assign the value to a temporary variable before passing it along to `conditional`.

```ts
// {#if expr; as alias}...{/if}
if (rf & 1) {
  ɵɵtemplate(0, App_Conditional_0_Template, 1, 0);
}
if (rf & 2) {
  let App_contFlowTmp;
  ɵɵconditional(0, (App_contFlowTmp = ctx.expr) ? 0 : -1, App_contFlowTmp);
}
```

### test(core): remove manually-written control flow instructions
Updates the control flow tests to use the compiler instead of manually-written instructions. Also adds a couple of tests that I was using along the way to verify that things work as expected.

crisbeto committed:
refactor(compiler): generate switch block instructions

Adds the logic to generate the instructions for `switch` instructions. For the following block:

```html
{#switch value()}
  {:case 0} case 0
  {:case 1} case 1
  {:case 2} case 2
  {:default} default
{/switch}
```

The compiler will produce the following output:

```ts
function App_Template(rf, ctx) {
  if (rf & 1) {
    ɵɵtemplate(0, App_Case_0_Template, 1, 0);
    ɵɵtemplate(1, App_Case_1_Template, 1, 0);
    ɵɵtemplate(2, App_Case_2_Template, 1, 0);
    ɵɵtemplate(3, App_Case_3_Template, 1, 0);
  }
  if (rf & 2) {
    let App_contFlowTmp;
    ɵɵconditional(0, (App_contFlowTmp = ctx.value()) === 0 ? 0 : App_contFlowTmp === 1 ? 1 : App_contFlowTmp === 2 ? 2 : 3);
  }
}
```

crisbeto committed:
refactor(compiler): generate if block instructions

Adds the logic to generate the instructions for `if` blocks. There are two primary use cases we need to account for:

A conditional that doesn't use the `as` parameter of the `if` block. To support it we generate a nested ternary expression that evaluates to the index of the template whose condition is truthy. If the block doesn't have an `else` branch, we pass in a special `-1` value which means that no view will be rendered.

Example with an `else`:
```ts
// {#if expr}
//   ...
//   {:else if otherExpr} ...
//   {:else} ...
// {/if}

if (rf & 1) {
  ɵɵtemplate(0, App_Conditional_0_Template, 0, 0);
  ɵɵtemplate(1, App_Conditional_1_Template, 0, 0);
  ɵɵtemplate(2, App_Conditional_2_Template, 0, 0);
}
if (rf & 2) {
  ɵɵconditional(0, ctx.expr ? 0 : ctx.otherExpr ? 1 : 2);
}
```

Example without an `else`:
```ts
// {#if expr}
//   ...
//   {:else if otherExpr} ...
// {/if}

if (rf & 1) {
  ɵɵtemplate(0, App_Conditional_0_Template, 0, 0);
  ɵɵtemplate(1, App_Conditional_1_Template, 0, 0);
}
if (rf & 2) {
  ɵɵconditional(0, ctx.expr ? 0 : ctx.otherExpr ? 1 : -1);
}
```

If a conditional captures it's value in an alias (e.g. `{#if expr; as foo}`) we need to assign the value to a temporary variable before passing it along to `conditional`.

```ts
// {#if expr; as alias}...{/if}
if (rf & 1) {
  ɵɵtemplate(0, App_Conditional_0_Template, 1, 0);
}
if (rf & 2) {
  let App_contFlowTmp;
  ɵɵconditional(0, (App_contFlowTmp = ctx.expr) ? 0 : -1, App_contFlowTmp);
}
```

crisbeto committed:
test(core): remove manually-written control flow instructions

Updates the control flow tests to use the compiler instead of manually-written instructions. Also adds a couple of tests that I was using along the way to verify that things work as expected.

iabu94 commented:
Thanks mates for the most awaited changes

pkozlowski-opensource commented:
> If I understand correctly, we'd have a behavior change with the current ngSwitch directive which used loose equality.
>  +1 for a strict equality 👍🏻

@JeanMeche correct, this would be behaviour change and it is not clear yet how breaky it would be for code migrated to the new syntax. Because of this we might have some sort of compatibility mode.

pkozlowski-opensource commented on packages/core/src/render3/instructions/control_flow.ts:
```diff
@@ -24,9 +24,11 @@ import {addLViewToLContainer, createAndRenderEmbeddedLView, getLViewFromLContain
  */
-export function ɵɵconditional<T>(containerIndex: number, matchingTemplateIndex: number, value?: T) {
+export function ɵɵconditional<T>(
+    containerIndex: number, matchingTemplateIndex: number, context?: T) {
```
I see from where you are coming with this change but I would prefer to keep it as-is. The reasoning being that it should be up to runtime to decide on the shape of the context. In this sense we are only interested in the value passed to the instruction.

pkozlowski-opensource commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_control_flow/if_context_reset_template.js:
```diff
@@ -0,0 +1,25 @@
+  }
+  if (rf & 2) {
+    let MyApp_contFlowTmp;
+    $r3$.ɵɵconditional(0, (MyApp_contFlowTmp = $r3$.ɵɵpureFunction1(5, _c0, ctx.expr1)).expr1Alias ? 0 : (MyApp_contFlowTmp = undefined) || ctx.expr2 ? 1 : (MyApp_contFlowTmp = $r3$.ɵɵpureFunction1(7, _c1, ctx.expr3)).expr3Alias ? 2 : (MyApp_contFlowTmp = undefined) || ctx.expr4 ? 3 : ctx.expr5 ? 4 : 5, MyApp_contFlowTmp);
```
As discussed yesterday, we don't need pure functions here

pkozlowski-opensource reviewed (commented):
This is great, almost there! Just added comments based on what we've discussed in person yesterday.

crisbeto commented:
@pkozlowski-opensource I've reworked the compiled output based on the feedback and addressed the remaining comments, PTAL. The one missing piece is throwing an error if `as` is used on an `else if` block. I have it in a separate commit in https://github.com/angular/angular/pull/51398.

pkozlowski-opensource reviewed (approved):
LGTM for the runtime changes and generated code 👍

AndrewKushnir commented:
This PR was merged into the repository by commit 9152de138727b25f5e194991cab5b658473a8582.
